src/webgpu/api/validation/copyBetweenLinearDataAndTexture.spec.ts

```diff
@@ -0,0 +1,745 @@
+      .combine(poptions('copyDepth', [0, 1]))
+  )
+  .fn(async t => {
+    const { copyWidth, copyHeight, copyDepth, method } = t.params;
+
+    const texture = t.device.createTexture({
+      size: { width: 2, height: 1, depth: 1 },
+      mipLevelCount: 1,
+      dimension: '1d',
+      format: 'rgba8unorm',
+      usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.COPY_DST,
+    });
+
+    // For 1d textures we require copyHeight and copyDepth to be 1,
+    // copyHeight or copyDepth being 0 should cause a validation error.
+    const _success = copyHeight == 1 && copyDepth == 1;
+
+    t.expectValidationError(() => {
+      t.testRun({
+        texture: texture,
+        origin: { x: 0, y: 0, z: 0 },
+        size: { width: copyWidth, height: copyHeight, depth: copyDepth },
+        mipLevel: 0,
+        bytesPerRow: 256,
+        rowsPerImage: 16,
+        dataSize: 16,
+        method: method,
+      });
+    }, !_success);
+  });
+
+enum Coord {
+  x = 'x',
+  y = 'y',
+  z = 'z',
+}
+
+const kCoords = [Coord.x, Coord.y, Coord.z];
+
+const kTestValuesForDivisibilityBy4 = [1, 2, 3, 4, 6, 8, 12];
+
+g.test('texel_block_alignemnts_on_origin')
+  .params(
+    params()
+      .combine(poptions('method', [TestMethod.WriteTexture, TestMethod.CopyBufferToTexture, TestMethod.CopyTextureToBuffer]))
+      .combine(poptions('value', kTestValuesForDivisibilityBy4))
+      .combine(poptions('coordinate_to_test', kCoords))
+      .combine(poptions('format', kTextureFormats))
+      .unless(({ format }) => !kTextureFormatInfo[format].copyable)
+  )
+  .fn(async t => {
+    const { value, coordinate_to_test, format, method } = t.params;
+
+    const texture = t.device.createTexture({
+      size: { width: 12, height: 12, depth: 12 },
+      mipLevelCount: 1,
+      format: format,
+      usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.COPY_DST,
+    });
+
+    let origin = { x: 0, y: 0, z: 0 };
+    let size = { width: 0, height: 0, depth: 0 };
+    let success = true;
+
+    switch (coordinate_to_test) {
+      case Coord.x: {
+       origin.x = value;
+       success = origin.x % kTextureFormatInfo[format].blockWidth === 0;
+       break;
+      }
+      case Coord.y: {
+       origin.y = value;
+       success = origin.y % kTextureFormatInfo[format].blockHeight === 0;
+       break;
+      }
+      case Coord.z: {
+       origin.z = value;
+       break;
+      }
+    }
+
+    t.expectValidationError(() => {
+      t.testRun({
+        texture: texture,
+        origin: origin,
+        size: size,
+        bytesPerRow: 1024,
+        rowsPerImage: 16,
+        dataSize: 1,
+        method: method,
+      });
+    }, !success);
+  });
+
+g.test('texel_block_alignemnts_on_size')
+  .params(
+    params()
+      .combine(poptions('method', [TestMethod.WriteTexture, TestMethod.CopyBufferToTexture, TestMethod.CopyTextureToBuffer]))
+      .combine(poptions('value', kTestValuesForDivisibilityBy4))
+      .combine(poptions('coordinate_to_test', kCoords))
```

Comment:
could similarly be `['width', 'height', 'depth']`. Or even `[0, 1, 2]` if you use the array overload of GPUExtent3D (kind of hard to read, though I think it would work well for texture_range_conditions.

Reply:
Done